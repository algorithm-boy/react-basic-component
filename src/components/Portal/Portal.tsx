import * as ReactDOM from 'react-dom';
import React, { SFC, useEffect } from 'react';
import { GeneralProps } from '../global.d';

export interface PortalProps extends GeneralProps {
  selector?: string;
  visible?: boolean;
  render?: () => React.ReactNode;
  onMount?: () => any;
  onUnmount?: () => any;
};

function getContainer (selector: string) {
  if (typeof window === 'undefined') { return; }
  const node = document.querySelector(selector) || document.body;
  return node;
}

const Portal: SFC<PortalProps> = props => {
  const { className, style = {}, children, selector = 'body', visible, render, onMount, onUnmount } = props;

  const container = getContainer(selector);
  const content = render ? render() : children;
  if (!container) return <>{ content }</>;

  /* eslint-disable */
  useEffect(() => {
    onMount && onMount();

    return () => onUnmount && onUnmount();
  }, [])
  /* eslint-enable */

  return ReactDOM.createPortal(
    <div
      className={className}
      style={{
        display: visible ? 'block' : 'none',
        ...style
      }}
    >
      { content }
    </div>,
    container
  );
};

export default Portal;
