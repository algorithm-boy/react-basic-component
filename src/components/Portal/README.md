# Portal(传送门)

## API

| 参数(props) | 说明(specification) | 类型(type) | 默认值(defaultProps) |
| --- | --- | --- | --- |
| className(optional) | Portal的类名 (The class property for Portal) | string | - |
| style(optional) | Portal的行内样式 (The inline style for Portal) | object | {} |
| selector(optional) | 将Portal传送至某个DOM的选择器，基于 `querySelector` API实现 (The DOM selector that Portal will Transfer to which based on `querySelector` API) | string | 'body' |
| visible(optional) | Portal是否显示的开关 (The switch for whether or not display Portal) | boolean | - |
| render(optional) | RenderProps，返回需要渲染的React组件，优先级高于children (RenderProps, which returns the rendered React Component with priority over children) | () => React.ReactNode | - |
| onMount(optional) | 挂载完毕的回调函数 (The callback when mount) | () => any | - |
| onUnmount(optional) | 卸载的回调函数 (The callback when unmount) | () => any | - |

## Example

```javascript
<Portal
  visbale
>
  Message
</Portal>
```

