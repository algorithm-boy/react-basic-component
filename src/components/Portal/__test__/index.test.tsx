import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Portal from '../index';

configure({ adapter: new Adapter() });

describe('Portal', () => {
  it('renders correctly - shallow', () => {
    const wrapper = shallow(<Portal visible><span>Text</span></Portal>);
    expect(wrapper.find('div').exists()).toBe(true);
    expect(wrapper.find('span').exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - mount', () => {
    const wrapper = mount(<Portal visible><span>Text</span></Portal>);
    expect(wrapper.contains(<Portal visible><span>Text</span></Portal>)).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it('pass visible parameter', () => {
    const wrapper = mount(<Portal visible><span>Text</span></Portal>);

    expect(wrapper.props().visible).toEqual(true);
  });

  it('pass selector parameter', () => {
    const wrapper = mount(<Portal visible selector='#div'><span>Text</span></Portal>);

    expect(wrapper.props().selector).toEqual('#div');
  });

  it('pass render parameter', () => {
    const wrapper = mount(<Portal visible render={() => <b>TextB</b>}><span>Text</span></Portal>);

    expect(wrapper.props().render).toBeTruthy();
    expect(wrapper.find('b').exists()).toBe(true);
    expect(wrapper.find('span').exists()).toBe(false);
  });

  it('pass onMount and onUnmount parameter', () => {
    const wrapper = mount(<Portal
        visible 
        onMount={() => console.info('onMount')}
        onUnmount={() => console.info('onUnmount')}
      ><span>Text</span></Portal>
    );

    expect(wrapper.props().onMount).toBeTruthy();
    expect(wrapper.props().onUnmount).toBeTruthy();
  });
});
