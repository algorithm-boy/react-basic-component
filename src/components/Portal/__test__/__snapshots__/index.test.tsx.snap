// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal renders correctly - mount 1`] = `ReactWrapper {}`;

exports[`Portal renders correctly - shallow 1`] = `ShallowWrapper {}`;
